src/visualizers/histogram.rs

```diff
@@ -0,0 +1,157 @@
+use super::{FillFrom, FillModifiers, RangeModifiers};
+use crate::utils::{HistogramBuffer, VisualizerBuffer, ValueScaling};
+
+use nih_plug_vizia::vizia::{prelude::*, vg};
+use std::sync::{Arc, Mutex};
+
+/// Real-time histogram displaying information that is stored inside a [`HistogramBuffer`]
+///
+/// # Example
+///
+/// Here's how to set up a histogram. For this example, you'll need a
+/// [`HistogramBuffer`](crate::utils::HistogramBuffer) to store your histogram information.
+///
+/// ```
+/// Histogram::new(cx, Data::histogram_buffer, (-32.0, 8.0), 0.1)
+///     .color(Color::rgba(0, 0, 0, 160))
+///     .background_color(Color::rgba(0, 0, 0, 60));
+/// ```
+///
+/// The histogram displays the range from -32.0dB to 8dB.
+/// it decays as 0.1 TODO, and a stroke and fill (background) color is provided.
+pub struct Histogram<L>
+where
+    L: Lens<Target = Arc<Mutex<HistogramBuffer>>>,
+{
+    buffer: L,
+    range: (f32, f32),
+}
+
+enum HistogramEvents {
+    UpdateRange((f32, f32)),
+    // UpdateDecay(f32),
+}
+
+impl<L> Histogram<L>
+where
+    L: Lens<Target = Arc<Mutex<HistogramBuffer>>>,
+{
+    pub fn new(
+        cx: &mut Context,
+        buffer: L,
+        range: impl Res<(f32, f32)> + Clone,
+    ) -> Handle<Self> {
+        Self {
+            buffer,
+            range: range.get_val(cx),
+        }
+        .build(cx, |_| {})
+        .range(range)
+    }
+}
+
+impl<L> View for Histogram<L>
+where
+    L: Lens<Target = Arc<Mutex<HistogramBuffer>>>,
+{
+    fn element(&self) -> Option<&'static str> {
+        Some("histogram")
+    }
+    fn event(&mut self, _cx: &mut EventContext, event: &mut Event) {
+        event.map(|e, _| match e {
+            HistogramEvents::UpdateRange(v) => self.range = *v,
+            // HistogramEvents::UpdateDecay(s) => self.decay = *s,
+        });
+    }
+    fn draw(&self, cx: &mut DrawContext, canvas: &mut Canvas) {
+        let bounds = cx.bounds();
+
+        let line_width = cx.scale_factor();
+
+        let x = bounds.x;
+        let y = bounds.y;
+        let w = bounds.w;
+        let h = bounds.h;
+
+        let mut stroke = vg::Path::new();
+        let binding = self.buffer.get(cx);
+        let bins = &(binding.lock().unwrap());
+        let nr_bins = bins.len();
+
+        // scale so the largest value becomes 1.
+        // let largest = self.buffer.iter().fold(std::f32::MIN, |a,b| a.max(*b));
+        let mut largest = 0.0;
+
+        for i in 0..nr_bins {
```

Comment:
How can I turn this into an iter?
Would that improve performance?

Reply:
I don't think you really need to do that. It would also mean creating a new iterator struct for the histogram buffer and implementing iter() or into_iter() for the buffer. It wouldn't really have any effect on performance, I think.

Reply:
I don't think you really need to do that. It would also mean creating a new iterator struct for the histogram buffer and implementing iter() or into_iter() for the buffer. It wouldn't really have any effect on performance, I think.